src/data-sources/components/DataSourceForm.tsx

```diff
@@ -70,15 +71,45 @@ const DataSourceForm = ( { children, onSave }: DataSourceFormProps ) => {
 	const steps = Children.toArray( children );
 	const singleStep = steps.length === 1 || screen === 'editDataSource';
 
-	const stepHeadings = [ 'Setup', 'Scope' ];
+	const getStepHeadings = () => {
+		const headings = [ __( 'Setup' ) ]; // First step is always Setup
+
+		if (
+			Children.toArray( children ).some(
+				child =>
+					isValidElement( child ) &&
+					( child.type === DataSourceForm.Scope || child.type === 'Scope' )
+			)
+		) {
+			headings.push( __( 'Scope' ) );
+		}
+
+		if (
+			Children.toArray( children ).some(
+				child =>
+					isValidElement( child ) &&
+					( child.type === DataSourceForm.Blocks || child.type === 'Blocks' )
+			)
+		) {
+			headings.push( __( 'Blocks' ) );
+		}
+
+		return headings;
+	};
+
+	const stepHeadings = getStepHeadings();
 
 	const canProceedToNextStep = () => {
 		const step = steps[ currentStep - 1 ];
 		if ( ! isValidElement< { canProceed?: boolean; displayName: string; uuid: string } >( step ) ) {
 			return false;
 		}
 		const { canProceed, displayName, uuid } = step.props;
-		return Boolean( canProceed ) && ( currentStep !== 1 || canUseDisplayName( displayName, uuid ) );
+		return (
+			( Boolean( canProceed ) &&
+				( currentStep !== 1 || canUseDisplayName( displayName, uuid ) ) ) ||
+			step.type === DataSourceForm.Blocks
+		);
```

Comment:
Getting a little gnarly, I find it helpful to split out complex boolean logic into semantic variables (u can probably think of better names).
```suggestion
		const isBlocksStep = step.type === DataSourceForm.Blocks;
		const noConflicts = Boolean( canProceed ) &&
			( currentStep !== 1 || canUseDisplayName( displayName, uuid ) );
		
		return isBlocksStep || noConflicts;
```